src/atomics/relaxed.md

```diff
@@ -0,0 +1,423 @@
+the old and new values — instead, it immediately updates, resulting in a diagram
+that look a bit more like this:
+
+```text
+Thread 1     data
+╭───────╮   ┌────┐
+│  = 1  ├─┐ │  0 │
+╰───────╯ │ └────┘
+          └─┬────┐
+            │  1 │
+            └────┘
+```
+
+We have now established a _modification order_ for `data`: a total, ordered list
+of distinct, separated values that it takes over its lifetime.
+
+On the loading side, we also obtain one additional ability: when there are
+multiple possible values to choose from in the modification order, instead of it
+triggering UB, exactly one (but it is unspecified which) value is chosen. This
+means that there are now _two_ potential executions of our program, with no way
+for us to control which one occurs:
+
+```text
+     Possible Execution 1       ┃       Possible Execution 2
+                                ┃
+Thread 1     data    Thread 2   ┃  Thread 1     data    Thread 2
+╭───────╮   ┌────┐   ╭───────╮  ┃  ╭───────╮   ┌────┐   ╭───────╮
+│ store ├─┐ │  0 ├───┤  load │  ┃  │ store ├─┐ │  0 │ ┌─┤  load │
+╰───────╯ │ └────┘   ╰───────╯  ┃  ╰───────╯ │ └────┘ │ ╰───────╯
+          └─┬────┐              ┃            └─┬────┐ │
+            │  1 │              ┃              │  1 ├─┘
+            └────┘              ┃              └────┘
+```
+
+Note that **both sides must be atomic to avoid the data race**: if only the
+writing side used atomic operations, the reading side would still have multiple
+values to choose from (UB), and if only the reading side used atomic operations
+it could end up reading the garbage data “in-between” `0` and `1` (also UB).
+
+> **NOTE:** This description of why both sides are needed to be atomic
+> operations, while neat and intuitive, is not strictly correct: in reality the
+> answer is simply “because the spec says so”. However, it is isomorphic to the
+> real rules, so it can aid in understanding.
+
+## Read-modify-write operations
+
+Loads and stores are pretty neat in avoiding data races, but you can’t get very
+far with them. For example, suppose you wanted to implement a global shared
+counter that can be used to assign unique IDs to objects. Naïvely, you might try
+to write code like this:
+
+```rust
+# use std::sync::atomic::{self, AtomicU64};
+static COUNTER: AtomicU64 = AtomicU64::new(0);
+pub fn get_id() -> u64 {
+    let value = COUNTER.load(atomic::Ordering::Relaxed);
+    COUNTER.store(value + 1, atomic::Ordering::Relaxed);
+    value
+}
+```
+
+But then calling that function from multiple threads opens you up to an
+execution like below that results in two threads obtaining the same ID:
+
+```text
+Thread 1   COUNTER   Thread 2
+╭───────╮   ┌───┐   ╭───────╮
+│ load  ├───┤ 0 ├───┤  load │
+╰───╥───╯   └───┘   ╰────╥──╯
+╭───⇓───╮ ┌─┬───┐   ╭────⇓──╮
+│ store ├─┘ │ 1 │ ┌─┤ store │
+╰───────╯   └───┘ │ ╰───────╯
+            ┌───┬─┘
+            │ 1 │
+            └───┘
```

Comment:
Maybe leave a note that the fact that `1` appears twice sequentially in the MO is intentional